[NVPTX] instcombine known pointer AS checks.

Artem-B opened the pull request:
This avoids crashing on impossible address space casts guarded by `__isGlobal/__isShared`.

Partially fixes https://github.com/llvm/llvm-project/issues/112760

It's still possible to trigger the issue by using explicit AS casts w/o AS checks, but LLVM should no longer crash on valid code.

Artem-B committed:
[NVPTX] instcombine known pointer AS checks.

This avoids crashing on impossible address space casts guarded by `__isGlobal/__isShared`.

AlexMaclean commented on llvm/include/llvm/Support/NVPTXAddrSpace.h:
```diff
@@ -0,0 +1,33 @@
+//===---------------- AMDGPUAddrSpace.h -------------------------*- C++ -*-===//
```
Looks like there are still some references to AMDGPU in here

Artem-B replied:
Good catch. Fixed.

Artem-B committed:
Clean up names in comments.

AlexMaclean commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6365,6 +6367,34 @@ static Value *simplifyUnaryIntrinsic(Function *F, Value *Op0,
+    auto *Ty = F->getReturnType();
+    unsigned AS = Op0->getType()->getPointerAddressSpace();
+    if (AS == NVPTXAS::ADDRESS_SPACE_GENERIC) {
+      if (auto *ASC = dyn_cast<AddrSpaceCastInst>(Op0))
```
Can AddrSpaceCastOperator be used here?

Artem-B replied:
Fixed.

Artem-B committed:
Cast directly to AddrSpaceCastOperator

Artem-B commented:
@nikic Are you OK with these changes?

AlexMaclean commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6365,6 +6367,32 @@ static Value *simplifyUnaryIntrinsic(Function *F, Value *Op0,
+  case Intrinsic::nvvm_isspacep_const: {
+    auto *Ty = F->getReturnType();
+    unsigned AS = Op0->getType()->getPointerAddressSpace();
+    if (AS == NVPTXAS::ADDRESS_SPACE_GENERIC) {
+      if (auto *ASC = dyn_cast<AddrSpaceCastInst>(Op0))
+        AS = ASC->getSrcAddressSpace();
+      else if (auto *ASCO = dyn_cast<AddrSpaceCastOperator>(Op0))
+        AS = ASCO->getOperand(0)->getType()->getPointerAddressSpace();
+    }
```
```suggestion
    if (AS == NVPTXAS::ADDRESS_SPACE_GENERIC)
      if (auto *ASCO = dyn_cast<AddrSpaceCastOperator>(Op0))
        AS = ASCO->getSrcAddressSpace();
```

Artem-B replied:
I believe we do want to handle ASC instruction, too, not just ASC operators. What am I missing here?

AlexMaclean replied:
https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/Operator.h#L30-L31

ASC operators include ASC instructions I believe.

Artem-B replied:
Ooh. This is convenient. Thanks for pointing it out. 

Coincidentally, the current tests in the patch didn't actually test handling of ASC instruction, as instcombine collapsed ASC(constant) into an ASC operator before we've got to handle the intrinsics. I'll update the code and the tests.

AlexMaclean commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6365,6 +6367,32 @@ static Value *simplifyUnaryIntrinsic(Function *F, Value *Op0,
+    if (AS == NVPTXAS::ADDRESS_SPACE_GENERIC ||
+        AS == NVPTXAS::ADDRESS_SPACE_PARAM)
+      return nullptr; // Got to check at run-time.
+    bool ASMatches = (AS == NVPTXAS::ADDRESS_SPACE_GLOBAL &&
```
This seems correct, but I think we can also do some simplification for `nvvm_isspacep_shared_cluster`. If the ptr type is non shared we know this check will return false.

Artem-B replied:
Will do.

Artem-B replied:
Done.

AlexMaclean reviewed (approved):
FWIW, this looks good to me baring some minor nits.

goldsteinn commented:
AFAICT this is the only target specific handling in InstSimplify which, at least as i understand, we attempt to keep as target independent as possible in a similar way to InstCombine because its for generalized middle end canonicalization.

If there is a need for this sort of change, I think it should be done more similarly to how we do `targetInstCombineIntrinsic` with InstCombine instead ofadding all the target intrinsics directly to the InstructionSimplify file.

Artem-B commented:
>  I think it should be done more similarly to how we do targetInstCombineIntrinsic

This, indeed, sounds like a better place to handle these intrinsics. Will do.

Artem-B committed:
Move isspacep intrinsics processing to NVPTX backend

Artem-B committed:
Use AS-specific pointers for ASC instruction testing

If we use a pointer to a constant, instcombine collapses it into an ASC operator.

Artem-B commented:
> > I think it should be done more similarly to how we do targetInstCombineIntrinsic
> 
> This, indeed, sounds like a better place to handle these intrinsics. Will do.

Done.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXTargetTransformInfo.cpp:
```diff
@@ -413,11 +416,65 @@ static Instruction *simplifyNvvmIntrinsic(IntrinsicInst *II, InstCombiner &IC) {
 
+// Returns an instruction pointer (may be nullptr if we do not know the answer).
+// Returns nullopt if `II` is not one of the `isspacep` intrinsics.
+static std::optional<Instruction *>
```
Can we just return nullptr and avoid the need for the `std::optional`?

Artem-B replied:
We have three outcomes here and the code is structured to convey that via the use of optional.
- not an isspacep intrinsic, and hence we have no answer -> nullopt. 
- if's an isspacep intrinsic, we do produce an answer, and it may be:
  - we've changed something, return non-null pointer
  - the answer is "we don't know", in which case we return nullptr

While we could return `nullptr` we'd continue checking for other intrinsics. It would happen to work, because nothing else handles the isspacep intrinsics, but I'd prefer the code to match the intent, and we do not intend to fall through into the code that deals with FP intrinsics.

AlexMaclean replied:
Fair enough.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXTargetTransformInfo.cpp:
```diff
@@ -413,11 +416,65 @@ static Instruction *simplifyNvvmIntrinsic(IntrinsicInst *II, InstCombiner &IC) {
+    auto *Ty = II.getType();
+    unsigned AS = Op0->getType()->getPointerAddressSpace();
+    // Peek through ASC to generic AS.
+    // TODO: we could dig deeper through both ASCs and GEPs.
```
Perhaps we should add this to rewriteIntrinsicWithAddressSpace as well?

Artem-B replied:
Do you mean that we should fill in the implementation of `rewriteIntrinsicWithAddressSpace` in general? 
Probably, but that should probably be done in a separate patch.

AlexMaclean replied:
I specifically mean that this same logic should also be happening in `rewriteIntrinsicWithAddressSpace` when we propagate a new AS into one of these isspacep intrinsics. I agree it can be in a separate patch though.

Artem-B committed:
revert changes to InstructionSimplify.cpp
